### PR TITLE
Convert test_flows.sh to python (part 2)

### PR DESCRIPTION
**Pull request by arnout**
More tests converted to python, with additional helpers.

### CODE REVIEW

File: tests/send_CAPI_command.py [hunk 1]
```diff
@@ -15,6 +15,54 @@ class CAPIReply(Enum):
...
+
+        value : str
+            The TLV value, as a string. It must be formatted according to the UCC rules, i.e. with
+            curly braces and hexadecimal numbers.
+        """
+        self.type = type_
+        self.length = length
+        self.value = value
+
+    def format(self, tlv_num: int = 0) -> str:
```
Comment (ghost) [inline tests/send_CAPI_command.py:41]: for ease of debugging, it could be useful to rename this function `__str__` (or `__repr__`), so you can just to `print(tlv)`.

File: tests/test_flows.py [hunk 2]
```diff
@@ -225,6 +295,82 @@ def test_initial_ap_config(self):
...
+            tlv(0x0F, 0x0001, "{0x00}"),
+            tlv(0x10, 0x0001, "{0x00}"))
+
+        time.sleep(3)
+        self.check_log(self.repeater1, "agent_wlan0", r".* tear down radio")
+        conn_map = self.get_conn_map()
+        repeater1 = conn_map[self.mac_repeater1]
+        repeater1_wlan0 = repeater1.radios[self.mac_repeater1_wlan0]
+        for vap in repeater1_wlan0.vaps.values():
+            if vap.ssid != b'N/A':
```
Comment (ghost) [inline tests/test_flows.py:370]: Nitpick: `SSID`s should probably be strings, not `bytes`, since they can contain unicode

**Review by ghost — APPROVED** [review]
Definitively an improvement :)

Some more nitpicks:
* Consider making the tlv class a named tuple (or a subclass of one). This kind-of conflicts with my earlier suggestion (about changing `tlv.format` to `tlv.__str__`) though, so do whichever one you like
* I don't like `send_CAPI_command.py` as a filename, I'd rather have something like `base_classes.py` (or simply `base.py`). Or maybe call it `capi.py` and put the tlv class in a seperate file

File: tests/send_CAPI_command.py (hunk 1 above)
Comment (arnout) [reply tests/send_CAPI_command.py:41]: This format function does something different than simply make a string representation, so NACK. I could add a __str__ that prints the three members though.

File: tests/test_flows.py (hunk 2 above)
Comment (arnout) [reply tests/test_flows.py:370]: SSIDs can contain any bytes, not limited to UTF-8. So it should most definitely be `bytes`.

**Comment by arnout** [comment]
>     * Consider making the tlv class a named tuple (or a subclass of one).

I don't see how a namedtuple is useful here... IMO it is way more awkward than a class.

I would love to use [attrs](https://github.com/hynek/attrs/) but that's unfortunately not a standard Python module so I wouldn't want to rely on it in the top-level test infra.

>     * I don't like `send_CAPI_command.py` as a filename, I'd rather have something like `base_classes.py` (or simply `base.py`). Or maybe call it `capi.py` and put the tlv class in a seperate file

I fully agree, but since send_CAPI_command.py is called from the existing test_flows.sh, I'd prefer to do that rename when test_flows.sh is removed.

**Comment by arnout** [comment]
I'll continue converting more tests, but it can be merged as is as well so ready for review.

File: tests/send_CAPI_command.py [hunk 3]
```diff
@@ -16,14 +16,63 @@ class CAPIReply(Enum):
...
+            starting from 1. `tlv_num` is that number. If there is only a single TLV, set it to
+            zero.
+
+        Returns
+        -------
+        str
+            The string representation of the TLV according to CAPI definition.
+        """
+        if tlv_num:
+            tlv_num = str(tlv_num)
```
Comment (LiorAmram) [inline tests/send_CAPI_command.py]: I think it's bad practice to change the type of a variable, especially when you explicitly mark it as int in the method signature. 
I would prefer defining a tlv_num_str variable.

File: tests/send_CAPI_command.py [hunk 4]
```diff
@@ -126,29 +193,20 @@ def dev_send_1905(self, dest: str, message_type: int, *tlvs: Tuple[int, int, str
...
-                tlv_num = str(tlv_num)
-            else:
-                tlv_num = ''
-            return "tlv_type{tlv_num},0x{tlv_type:02x}," \
-                   "tlv_length{tlv_num},0x{tlv_length:04x}," \
-                   "tlv_value{tlv_num},{tlv_value}".format(**locals())
-
-        cmd = "DEV_SEND_1905,DestALid,{dest:s},MessageTypeValue,0x{message_type:04x}".format(**locals())
+        cmd = "DEV_SEND_1905,DestALid,{dest:s},MessageTypeValue,0x{message_type:04x}"\
+            .format(locals())
```
Comment (LiorAmram) [inline tests/send_CAPI_command.py]: Why did you remove the unpacking? Now the format method receives a dict, and this code should fail...

File: tests/test_flows.py [hunk 5]
```diff
@@ -234,6 +234,33 @@ def test_initial_ap_config(self):
         self.check_log(self.repeater1, "agent_wlan0", r".* Controller configuration \(WSC M2 Encrypted Settings\)")
         self.check_log(self.repeater1, "agent_wlan2", r".* Controller configuration \(WSC M2 Encrypted Settings\)")
 
+    def test_ap_config_renew(self):
+        # Regression test: MAC address should be case insensitive
+        mac_repeater1_upper = self.mac_repeater1.upper()
+        # Configure the controller and send renew
+        self.gateway_ucc.cmd_reply("DEV_RESET_DEFAULT")
```
Comment (tomereli) [inline tests/test_flows.py:305]: Since autoconfiguration renew should be reused by multiple tests (4.4.3, 4.2.1, 4.8.1, etc.) - we should implement a helper function for ap_config_renew, preferable in a followup commit.
Something like:
```
def trigger_ap_config_renew(self, al_mac, credentials-2.4g, credentials-5g)
```

(nitpick) since this can be done in a separate PR, but I rather keep this comment open till its done.

File: tests/test_flows.py [hunk 6]
```diff
@@ -261,6 +261,37 @@ def test_ap_config_renew(self):
...
+            "DEV_SET_CONFIG,bss_info1,{self.mac_repeater1} 8x".format(self = self))
+        # Send renew message
+        self.gateway_ucc.dev_send_1905(self.mac_repeater1, 0x000A,
+            tlv(0x01, 0x0006, "{" + self.mac_gateway + "}"),
+            tlv(0x0F, 0x0001, "{0x00}"),
+            tlv(0x10, 0x0001, "{0x00}"))
+
+        time.sleep(3)
+        self.check_log(self.repeater1, "agent_wlan0", r".* tear down radio")
+        # TODO check that conn map was updated on controller
```
Comment (tomereli) [inline tests/test_flows.py]: Why do we care about the connection map? The TODO should say that we should trigger topology query and check the AP operational BSS TLV which should be empty for all radios.

File: tests/test_flows.py [hunk 7]
```diff
@@ -261,6 +261,37 @@ def test_ap_config_renew(self):
         if not bssid1:
             self.fail("repeater1 didn't configure Multi-AP-24G-1")
 
+    def test_ap_config_bss_tear_down(self):
+        # Configure the controller and send renew
```
Comment (tomereli) [inline tests/test_flows.py:329]: So this should already use a helper function.

**Review by LiorAmram — DISMISSED** [review]

File: tests/test_flows.py [hunk 8]
```diff
@@ -133,6 +162,31 @@ def docker_command(self, device: str, *command: str) -> bytes:
         '''Execute `command` in docker container `device` and return its output.'''
         return subprocess.check_output(("docker", "exec", device) + command)
 
+    def beerocks_command(self, command: str) -> bytes:
```
Comment (tomereli) [inline tests/test_flows.py]: `beerocks_cli_command` is more appropriate IMO (nitpick)

File: tests/test_flows.py [hunk 9]
```diff
@@ -277,7 +331,16 @@ def test_ap_config_bss_tear_down(self):
 
         self.check_log(self.repeater1, "agent_wlan0", r"Received credentials for ssid: Multi-AP-24G-3 .* bss_type: 2")
         self.check_log(self.repeater1, "agent_wlan2", r".* tear down radio")
-        # TODO check that conn map was updated on controller
+        conn_map = self.get_conn_map()
```
Comment (tomereli) [inline tests/test_flows.py:344]: So now I understand the conn map TODO comment, but still, we should imitate the exact steps which are tested in the testbed, so once #924 is available we should convert this to check the AP Operational BSS TLV instead of relying on the connection map.

File: tests/test_flows.py [hunk 10]
```diff
@@ -245,7 +245,7 @@ def check_log(self, device: str, program: str, regex: str) -> bool:
         '''Verify that on "device" the logfile for "program" matches "regex", fail if not.'''
         logfilename = os.path.join(self.rootdir, 'logs', device, 'beerocks_{}.log'.format(program))
         try:
-            # HACK check_log is often used immediately after triggering a message on the other side.
+            # HACK self.check_log(is, "often", rused immediately after triggering a message on the other side.)
```
Comment (tomereli) [inline tests/test_flows.py]: What's up with this comment?

File: tests/test_flows.py [hunk 11]
```diff
@@ -453,6 +453,46 @@ def test_combined_infra_metrics(self):
         self.check_log(self.repeater1, "agent", "Received TLV_TRANSMITTER_LINK_METRIC")
         self.check_log(self.repeater1, "agent", "Received TLV_RECEIVER_LINK_METRIC")
 
+    def test_client_capability_query(self):
+        sta_mac1 = '00:00:00:11:00:22'
+        sta_mac2 = '00:00:00:11:00:33'
+
+        self.gateway_ucc.dev_send_1905(self.mac_repeater1, 0x8009,
```
Comment (tomereli) [inline tests/test_flows.py:464]: A comment explaining the 1905 command being sent will help understand the flow better.

**Review by tomereli — APPROVED** [review]
mainly nitpick comments, but overall I can't really fault this. LGTM

File: tests/send_CAPI_command.py (hunk 3 above)
Comment (arnout) [reply tests/send_CAPI_command.py]: Good point. Done.

File: tests/send_CAPI_command.py (hunk 4 above)
Comment (arnout) [reply tests/send_CAPI_command.py]: That was an accident, fixed.

File: tests/test_flows.py (hunk 6 above)
Comment (arnout) [reply tests/test_flows.py]: There should be no need to trigger topology query, since the controller should do that automatically upon receiving a topology notification -> #925.

I added the TODO because I removed the corresponding code that existed in test_flows.sh, and then reimplemented it in a different way in the next commit. Since it wasn't a one-to-one mapping I thought it was better to do it in a separate commit, and I didn't want to bother with implementing it the original way first.

File: tests/test_flows.py (hunk 9 above)
Comment (arnout) [reply tests/test_flows.py:344]: I disagree that we should do the exact steps of the testbed.

We should *at least* do the same as the testbed, but I think it's definitely OK if we do more.

File: tests/test_flows.py (hunk 10 above)
Comment (arnout) [reply tests/test_flows.py]: Hahaha, search-replace effect!
